README
Signals-XSIG
------------

Signals::XSIG  extends Perl's simple %SIG interface to trapping
signals and installing multiple handlers for signals.

See the perldoc for a description and use cases of this module.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Signals::XSIG

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Signals-XSIG

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Signals-XSIG

    CPAN Ratings
        http://cpanratings.perl.org/d/Signals-XSIG

    Search CPAN
        http://search.cpan.org/dist/Signals-XSIG


LICENSE AND COPYRIGHT

Copyright (c) 2010-2013 Marty O'Brien

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Signals-XSIG

0.13    2013-01-14
        Backport to 5.6. Incorporate new defaults for gnukfreebsd and
        mirbsd systems.

0.12    2011-07-08
        Fix so $SIG{sig}=\&bogus, $SIG{sig}='bogus', $SIG{sig}=*bogus
        don't crash the program when the signal is handled.
        Doc and test for override DEFAULT behavior feature.

0.11    2011-04-18
        More test and doc tweaks.

0.10    2011-02-11
        Test tweak to help pass on OpenBSD, MSWin32.

0.09    2010-11-10
        Test changes. More tweaks to default behavior on bsd, irix systems.

0.08    2010-11-07
        Test changes -- when default test fails for some signal, run
        spike/analyze_default_signal_behavior.pl to include the correct
        default behavior in the test output.

0.07    2010-11-05
        Test and default signal handling fixes.

0.06    2010-11-03
        Test and doc fixes. 

0.05    2010-11-02
        Update default behavior for lots of platforms from CPAN test results.

0.04    2010-10-28
        Implement S::X::TieArray->{start} attribute to (1) save memory and
        (2) make it easier to distinguish "pre-handlers" and "post-handlers".

0.03    2010-10-24
        Removed %REGISTRY refs.

0.02    2010-10-08
        Introduces %XSIG.
        Added defaul signal behaviors for solaris.

0.01    Tie %SIG, use signal handler registry to implement multiple
        signal handlers.


t/SignalHandlerTest.pm
# t::SignalHandlerTest
# some global routines to help the testing of
# Signals::XSIG.
##################################################

package t::SignalHandlerTest;
use Exporter;
use Config;
use strict;
use warnings;
our @ISA = qw(Exporter);
our @EXPORT = qw(trigger appropriate_signals alias_pair);

my @no = split ' ', $Config{sig_num};
my @name = split ' ', $Config{sig_name};

# returns some signal names that are appropriate for testing
# on this platform
sub appropriate_signals {
  if (wantarray) {
    if ($^O eq "MSWin32") {
      return qw(INT QUIT);
    } else {
      return qw(USR1 USR2);
    }
  } else {
    if ($^O eq 'MSWin32') {
      return 'INT';
    } else {
      return 'USR1';
    }
  }
}

# returns a pair of aliased signal names (like CLD and CHLD)
# that is appropriate for this platform
sub alias_pair {
  my @seen = ();
  for (my $i=0; $i<@no; $i++) {
    if ($seen[$no[$i]]) {
      return ($seen[$no[$i]], $name[$i]);
    } else {
      $seen[$no[$i]] = $name[$i];
    }
  }
}

# raise a signal, trying to be platform-independent
sub trigger ($) {
  my ($signal) = @_;
  if ($^O eq "MSWin32" && $signal eq "ALRM") {
    if (0 && $Config{PERL_VERSION} <= 8) {

      # is there a way to trigger a signal handler,
      # any signal handler, on Strawberry 5.8?

      # sleep, alarm may be incompatible
      my $j=0;
      my $t = time + 2;
      alarm 1;
      while (time < $t) {
	$j += rand() - 0.5;
      }
    } elsif (eval { require Time::HiRes; Time::HiRes::alarm(0); 1 }) {
      Time::HiRes::alarm(0.25);
      Time::HiRes::sleep(0.75);
      Time::HiRes::alarm(0);
    } else {
      alarm 1;
      sleep 2;
      alarm 0;
    }
  } else {
    kill $signal, $$;
  }
}

1;


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;
use Config;

if ($Config{"PERL_VERSION"} == 8) {
  print <<'__END_WARNING__;';

*******************************************
*     !!!!!!!!!! WARNING !!!!!!!!!!!!!!   *
* In Perl v5.8, the construction          *
*                                         *
*      $tied_hash{$key} = *value          *
*                                         *
* causes a fatal error. This module works *
* by making %SIG into a tied hash, so     *
* code that used the somewhat esoteric    *
* assignment                              *
*                                         *
*      $SIG{signal} = *function;          *
*                                         *
* will not work with this module.         *
*******************************************

__END_WARNING__;
}

if (0 && $Config{"PERL_VERSION"} == 13) {
  # don't think we still have this problem in 0.03
  print <<'__END_WARNING__;';

****************************************
*    !!!!!!!!! WARNING !!!!!!!!!!!!    *
* Behavior of this module is different *
* under Perl 5.13 than under earlier   *
* versions. Consult the BUGS AND       *
* LIMITATIONS section of the perldoc   *
* for more details.                    *
****************************************

__END_WARNING__;
}

WriteMakefile(
    NAME                => 'Signals::XSIG',
    AUTHOR              => q{Marty O'Brien <mob@cpan.org>},
    VERSION_FROM        => 'lib/Signals/XSIG.pm',
    ABSTRACT_FROM       => 'lib/Signals/XSIG.pm',
    (eval $ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
    PREREQ_PM => {
        'Test::More' => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Signals-XSIG-*'
			   . ' *.stackdump default-data/*'
                           . ' control* experimental* out*' },
);


spike/analyze_default_signal_behavior.pl
# analyze_default_signal_behavior.pl: see what each signal does
# to a Perl program when the "DEFAULT" signal handler is set on 
# that program. The results can be appended to the
# lib/Signals/XSIG/Default.pm  file.

use IO::Handle;
use POSIX ':sys_wait_h';
use Config;
use strict;
use warnings;
$| = 1;

if (@ARGV < 2) {
  print STDERR qq[
This program will experimentally determine the default behavior 
of each signal on your system. The data collected will be
helpful in creating an appropriate  Signals/XSIG/Default.pm  
file.
];

}

my (@IGNORE, @SUSPEND, @TERMINATE, @UNKNOWN);
my $num_simultaneous = shift @ARGV || 10;
@IGNORE = ('__WARN__','__DIE__');

my @sigs = (sort keys %SIG, 'ZERO');
@sigs = @ARGV if @ARGV > 0;
my $abort_status;

if (@ARGV < 2) {
  printf STDERR "There are %d signals to analyze.\n", scalar @sigs;
  print STDERR "This may take a few minutes.\n\n";
}

# figure out the exit status of a program that calls POSIX::abort()
sub abort_status {
  use POSIX ();
  if (!defined $abort_status) {
    if (fork() == 0) {
      POSIX::abort();
      exit 0;
    }
    wait;
    $abort_status = $? || -9E9;
  }
  return $abort_status;
}

sub analysis_file {
  my ($signal) = @_;
  "siganal$signal.txt";
}

sub analyze_default_behavior_for_signal {
  my ($sig, $i, $analysis_file, $analysis_script) = @_;

  $analysis_file ||= analysis_file($sig);
  $analysis_script ||= "siganal$i.pl";
  unlink $analysis_file, $analysis_script;

  open my $cfh, '>', $analysis_script;
  print $cfh qq[

\$SIG{'$sig'} = 'DEFAULT';
my \$n = sleep 2;
my \$msg = "CHILD \$n / 4\n";
open(F, '>>', '$analysis_file');
print F \$msg;
close F;
exit 0;

];

  close $cfh;
  my $status = 'unknown';
  my ($pid, $win32ProcObj);

  if ($^O eq 'MSWin32') {
    require Win32::Process;
    require Win32;
    Win32::Process::Create( $win32ProcObj, 
			    $^X, 
			    "$^X $analysis_script",
			    0, 
			    &Win32::Process::NORMAL_PRIORITY_CLASS,
			    "." ) || die Win32::GetLastError();
    $pid = $win32ProcObj->GetProcessID();    
  } else {
    $pid = fork();
    if ($pid == 0) {
      %SIG = ();
      exec $^X, $analysis_script;
      die;
    }
  }

  sleep 2;
  my $nk = kill $sig, $pid;
  sleep 1;
  my $r = waitpid $pid, &WNOHANG;
  $status = $? if $r == $pid;
  sleep 5;
  $nk = kill 'CONT', $pid;
  sleep 1;
  $r = waitpid $pid, &WNOHANG;
  $status = $? if $r == $pid;
  $nk = kill 'KILL', $pid;
  $r = waitpid $pid, 0;
  $status = $? if $r == $pid;
  if (defined $win32ProcObj) {
    $win32ProcObj->Wait(&Win32::Process::INFINITE);
  }
  my $msg = "Status: $status\n";
  open my $fh, '>>', $analysis_file;
  print $fh $msg;
  close $fh;

  unlink $analysis_script;
  return $analysis_file;
}

$::j=0;
print "[$^O]\n";
while (@sigs) {
  my @sigz = splice @sigs, 0, $num_simultaneous;
  my $i = 0;
  foreach my $sig (@sigz) {
    if (fork() == 0) {
      analyze_default_behavior_for_signal($sig,$i);
      exit 0;
    }
    $i++;
  }
  wait foreach @sigz;
  foreach my $sig (@sigz) {
    parse_analysis_file($sig,analysis_file($sig));
  }
}

sub parse_analysis_file {
  my ($sig,$file) = @_;
  open G, '<', $file;
  my @g = <G>;
  close G;

  my $i = ++$::j;
  my @sig_name = split ' ', $Config{sig_name};
  my @sig_num = split ' ', $Config{sig_num};
  my ($sig_no) = grep { $sig_name[$_] eq $sig } 0..$#sig_num;
  $sig_no ||= 9999;
  $sig_no = $sig_num[$sig_no];
  $sig_no ||= '';

  my ($sleep_result, $sleep_benchmark) = $g[0] =~ /CHILD (\d+) \/ (\d+)/;
  if (defined $sleep_benchmark && $sleep_result > $sleep_benchmark) {
    # the program completed but took longer than ~4 seconds.
    # This means it was suspended and then resumed several seconds later.
    push @SUSPEND, $sig;
    printf STDERR "%d. SIG", $i;
    printf "%-7s [%s] => %s\n", $sig, $sig_no, "SUSPEND";
  } else {
    my ($status) = $g[-1] =~ /Status: (\d+)/;
    if ($status eq "0") {
      # The program completed normally in a regular amount of time.
      # The signal was ignored or not received.
      push @IGNORE, $sig;
      printf STDERR "%d. SIG", $i;
      printf "%-7s [%s] => %s\n", $sig, $sig_no, "IGNORE";
    } elsif ($status > 0) {
      # The program did not complete normally.
      # The signal terminated the program.
      push @TERMINATE, $sig;

      printf STDERR "%d. SIG", $i;
      if ($status == $sig_no << 8) {
	# exit status is divisible by 256. Like quitting with  exit()
	printf "%-7s [%s] => %s\n", $sig, $sig_no, "EXIT $sig_no";
      } elsif (0 && $status == abort_status()) {
	# exit status same as abort status (see &abort_status).
	printf "%-7s [%s] => %s\n", $sig, $sig_no, "ABORT";
      } else {
	# exit status not divisible by 256. Only way to do this
	# reliably is to actually raise the signal
	printf "%-7s [%s] => %s\n", $sig, $sig_no, "TERMINATE $status";
      }
    } else {
      push @UNKNOWN, $sig;
      printf STDERR "%d. SIG", $i;
      printf "%-7s [%s] => %d %s\n", $sig, $sig_no, $status, "UNKNOWN";
    }
  }
  unlink $file;
}


lib/Signals/XSIG/Default.pm
# emulate default behaviors for the various signals.

package Signals::XSIG::Default;

## no critic (RequireLocalizedPunctuationVars)

use strict;
use warnings;
use Config;
use Carp;
use POSIX ();
use Exporter;

our @ISA = qw(Exporter);
our @EXPORT = qw(%DEFAULT_BEHAVIOR);

our %DEFAULT_BEHAVIOR;
our $VERSION = '0.12';

my @snam = split ' ', $Config{sig_name};
my @snum = split ' ', $Config{sig_num};

sub import {
    my $ignore = 1;
    while (<DATA>) {
	next if /^#/;
	next unless /\S/;
	if (/^\[(.+)\]/) {
	    if ($1 eq 'default' || $1 eq $^O) {
		$ignore = 0;
	    } else {
		$ignore = 1;
	    }
	} elsif (/\{(.+)\/(\d+)\}/) {
	    if ($1 eq $^O && $2 <= int($] * 1000)) {
		$ignore = 0;
	    } else {
		$ignore = 1;
	    }
	} elsif (!$ignore) {
	    s/^\d+\. //;
	    s/^SIG//;
	    my ($sig, $num, $behavior) = /^(\w+)\s+\[(\d*)\]\s+=>\s+(.+)/;
	    if (defined $sig) {
		$DEFAULT_BEHAVIOR{$sig} = $behavior;
	    }
	}
    }
    return;
}

sub perform_default_behavior {
    my ($signal, @args) = @_;

#print "XXXXXX performing default behavior for $signal ...\n";

    my $funcname = 'default_SIG' . $signal;
    if (defined &$funcname) {
	no strict 'refs';                    ## no critic (NoStrict)
	return if $funcname->($signal, @args);
    }

    my $behavior = $DEFAULT_BEHAVIOR{$signal};
    if (!defined $behavior) {
	if ($signal =~ /^NUM(\d+)/) {
	    my $signum = 0 + $1;
	    $behavior = $DEFAULT_BEHAVIOR{"NUMxx"};
	    $behavior =~ s/xx/$signum/;
	}
	if (!defined $behavior) {
	    croak "Signals::XSIG: no default behavior is specified ",
	        "for SIG$signal. Terminating this program.\n";
	}
    }

    if (ref($behavior) eq 'CODE') {
	if (defined &$behavior) {
	    $behavior->($signal);
	    return;
	} else {
	    carp "Signals::XSIG: Default behavior for SIG$signal ",
	        "is not set to a valid subroutine.";
	    return;
	}
    }

    if ($behavior eq 'IGNORE') {
	return;
    }

    if ($behavior eq 'SUSPEND') {
	suspend($signal);
	# ... then wait for the SIGCONT ... 
	return;
    }

    if ($behavior =~ /^ABORT/) {
	untie %SIG;
	%SIG = ();
	$SIG{$signal} = $SIG{"ABRT"} = "DEFAULT";
	killprog_with_signal("ABRT");
	POSIX::abort();
	croak "Abort\n";
    }
    if ($behavior =~ /^SIGSEGV/) {
	killprog_with_signal('SEGV');
	croak "Abort\n";
    }

    if ($behavior =~ /^EXIT (\d+)/) {
	my $exit_code = $1;
	exit($exit_code);
    }

    if ($behavior =~ /^TERMINATE/) {
	my $number;
	for (my $i=0; $i<@snum; $i++) {
	    $number = $snum[$i] if $signal eq $snam[$i];
	}

	killprog_with_signal($signal, $number);
	croak "default behavior for SIG$signal should have killed script ",
            "but for some reason it didn't  :-(\n";
    }

    croak "Signals::XSIG: unknown behavior \"$behavior\" ",
        "for SIG$signal. Terminating this program.\n";
}

sub killprog_with_signal {
    my ($sig,$sig_no) = @_;
    untie %SIG;
    %SIG = ();
    $SIG{$sig} = 'DEFAULT';

    unless ($sig_no) {
	my @sig_name = split ' ', $Config{sig_name};
	($sig_no) = grep { $sig eq $sig_name[$_] } split ' ',$Config{sig_num};
    }

    kill $sig, $$;
    sleep 1 if $^O eq 'MSWin32';
    eval {
	use POSIX ();
	if ($sig_no) {
	    # this is needed for Linux
	    POSIX::sigaction($sig_no, &POSIX::SIG_DFL);
	    POSIX::sigprocmask(&POSIX::SIG_UNBLOCK, 
			       POSIX::SigSet->new($sig_no));
	}
    } or ();
    kill $sig, $$;
    sleep 1 if $^O eq 'MSWin32';

    my $miniprog = q[$SIG{'__SIGNAL__'}='DEFAULT';
                     kill '__SIGNAL__',$$;sleep 1+"MSWin32"eq$^O;die];
    $miniprog =~ s/__SIGNAL__/$sig/g;
    exec($^X, "-e", $miniprog);
}

# in principle, SIGSTOP cannot be trapped.
sub suspend {
    if ($^O eq 'MSWin32') {
	# MSWin32 doesn't have signals as such.
	# Win32::API->SuspendProcess / SuspendThread ?
	# Win32::Process->suspend ?
	# Win32::Thread->suspend ?
	if ($$ > 0) {
	    # suspend process
	    #   enumerate all threads in process
	    #   suspend each thread
	} else {
	    # suspend thread
	}
    }
    return kill 'STOP', $$;
}

##################################################################

# system specific and other special behaviors.
# Signals that don't fall into the terminate/suspend/ignore
# paradigm or that have other special needs can be implemented
# below.
# Return true if the signal is "handled" and no further 
# processing is necessary.

sub default_SIG__WARN__ {          ## no critic (Unpacking)
    CORE::warn @_;
    return 1;
}

sub default_SIG__DIE__ {          ## no critic (Unpacking)
    CORE::die @_;
    return 1;
}

1;

=head1 NAME

Signals::XSIG::Default

=head1 DESCRIPTION

Module for emulating the default behavior for all
signals in your system. The emulator is used when you have
used L<Signals::XSIG> to register more than one 
handler for a signal, and at least one of those 
handlers is C<DEFAULT>.

See L<Signals::XSIG> for much more information.

=cut



# see  spike/analyze_default_signal_behavior.pl

# for each new system that is available to us, run
#    spike/analyze_default_signal_behavior.pl
# and include that data at the end of this file ...
#
# we can also infer behavior from CPAN tester results,
# see t/20-defaults.t
#
__DATA__
[default]
ABRT [] => ABORT
ALRM [] => TERMINATE
BREAK [] => TERMINATE
BUS [] => TERMINATE
CHLD [] => IGNORE
CLD [] => IGNORE
CONT [] => IGNORE
EMT [] => TERMINATE
FPE [] => TERMINATE
HUP [] => TERMINATE
ILL [] => TERMINATE
INT [] => TERMINATE
IO [] => TERMINATE
IOT [] => TERMINATE
KILL [] => TERMINATE
LOST [] => TERMINATE
NUMxx [] => TERMINATE xx
PIPE [] => TERMINATE
POLL [] => TERMINATE
PROF [] => TERMINATE
PWR [] => TERMINATE
QUIT [] => TERMINATE
SEGV [] => ABORT
STKFLT [] => TERMINATE
STOP [] => SUSPEND
SYS [] => TERMINATE
TERM [] => TERMINATE
TRAP [] => TERMINATE
TSTP [] => SUSPEND
TTIN [] => SUSPEND
TTOU [] => SUSPEND
URG [] => IGNORE
USR1 [] => TERMINATE
USR2 [] => TERMINATE
VTALRM [] => TERMINATE
WINCH [] => IGNORE
XCPU [] => TERMINATE
XFSZ [] => TERMINATE
ZERO [] => IGNORE
__DIE__ [] => IGNORE
__WARN__ [] => IGNORE

[MSWin32]
ABRT    [22] => EXIT 22
ALRM    [14] => EXIT 14
BREAK   [21] => TERMINATE 21
CHLD    [20] => EXIT 20
CLD     [20] => EXIT 20
CONT    [25] => EXIT 25
FPE     [8] => EXIT 8
HUP     [1] => EXIT 1
ILL     [4] => EXIT 4
INT     [2] => IGNORE
KILL    [9] => EXIT 9
NUMxx   [] => EXIT xx
PIPE    [13] => EXIT 13
QUIT    [21] => IGNORE
SEGV    [11] => EXIT 11
STOP    [23] => EXIT 23
TERM    [15] => IGNORE
ZERO    [] => IGNORE
__DIE__ [] => IGNORE
__WARN__ [] => IGNORE

[linux]
ABRT    [6] => TERMINATE 6
ALRM    [14] => TERMINATE 14
BUS     [7] => TERMINATE 7
CHLD    [17] => IGNORE
CLD     [17] => IGNORE
CONT    [18] => IGNORE
FPE     [8] => TERMINATE 8
HUP     [1] => TERMINATE 1
ILL     [4] => TERMINATE 4
INT     [2] => TERMINATE 2
IO      [29] => TERMINATE 29
IOT     [6] => TERMINATE 6
KILL    [9] => TERMINATE 9
NUMxx   [] => TERMINATE xx
PIPE    [13] => TERMINATE 13
POLL    [29] => TERMINATE 29
PROF    [27] => TERMINATE 27
PWR     [30] => TERMINATE 30
QUIT    [3] => TERMINATE 3
RTMAX   [64] => TERMINATE 64
RTMIN   [34] => TERMINATE 34
SEGV    [11] => TERMINATE 11
STKFLT  [16] => TERMINATE 16
STOP    [19] => SUSPEND
SYS     [31] => TERMINATE 31
TERM    [15] => TERMINATE 15
TRAP    [5] => TERMINATE 5
TSTP    [20] => SUSPEND
TTIN    [21] => SUSPEND
TTOU    [22] => SUSPEND
UNUSED  [31] => TERMINATE 31
URG     [23] => IGNORE
USR1    [10] => TERMINATE 10
USR2    [12] => TERMINATE 12
VTALRM  [26] => TERMINATE 26
WINCH   [28] => IGNORE
XCPU    [24] => TERMINATE 24
XFSZ    [25] => TERMINATE 25
ZERO    [0] => IGNORE
__DIE__ [] => IGNORE
__WARN__ [] => IGNORE

[cygwin]
ABRT    [6] => TERMINATE 134
ALRM    [14] => TERMINATE 14
BUS     [10] => TERMINATE 10
CHLD    [20] => IGNORE
CLD     [20] => IGNORE
CONT    [19] => IGNORE
EMT     [7] => TERMINATE 7
FPE     [8] => TERMINATE 8
HUP     [1] => TERMINATE 1
ILL     [4] => TERMINATE 4
INT     [2] => TERMINATE 2
IO      [23] => IGNORE
KILL    [9] => TERMINATE 9
LOST    [29] => TERMINATE 29
PIPE    [13] => TERMINATE 13
POLL    [23] => IGNORE
PROF    [27] => TERMINATE 27
PWR     [29] => TERMINATE 29
QUIT    [3] => TERMINATE 131
RTMAX   [32] => TERMINATE 32
RTMIN   [32] => TERMINATE 32
SEGV    [11] => TERMINATE 11
STOP    [17] => SUSPEND
SYS     [12] => TERMINATE 12
TERM    [15] => TERMINATE 15
TRAP    [5] => TERMINATE 5
TSTP    [18] => SUSPEND
TTIN    [21] => SUSPEND
TTOU    [22] => SUSPEND
URG     [16] => IGNORE
USR1    [30] => TERMINATE 30
USR2    [31] => TERMINATE 31
VTALRM  [26] => TERMINATE 26
WINCH   [28] => IGNORE
XCPU    [24] => TERMINATE 24
XFSZ    [25] => TERMINATE 25
ZERO    [] => IGNORE
__DIE__ [] => IGNORE
__WARN__ [] => IGNORE

# www.cpantesters.org/cpan/report/4d60a820-e5c2-11df-bf62-5eb33ef6c52e
[darwin]
ABRT   [6] => TERMINATE 6
EMT    [7] => TERMINATE 7
SEGV   [11] => TERMINATE 11
IO     [] => IGNORE
IOT    [6] => TERMINATE 6

# www.cpantesters.org/cpan/report/4b99dd2a-e60f-11df-bb29-ad544afd17af
[dragonfly]
ABRT [6] => TERMINATE 134
EMT [7] => TERMINATE 135
SEGV [11] => TERMINATE 139
IO [] => IGNORE
IOT [] => TERMINATE 134

# www.cpantesters.org/cpan/report/62d0832e-e621-11df-858d-e879df34a846a
[freebsd]
ABRT [6] => TERMINATE 134
EMT [7] => TERMINATE 135
SEGV [11] => TERMINATE 139
IO [] => IGNORE
IOT [] => TERMINATE 134

[gnukfreebsd]
ABRT [6] => TERMINATE 134
EMT [7] => TERMINATE 135
SEGV [11] => TERMINATE 139
IO [] => IGNORE
IOT [] => TERMINATE 134

# www.cpantesters.org/cpan/report/c8635a72-e5d6-11df-a833-d9c7245fd73a
[irix]
ABRT [6] => TERMINATE 134
EMT [7] => TERMINATE 135
SEGV [11] => TERMINATE 139
RTMIN [49] => TERMINATE 49
RTMAX [64] => TERMINATE 64
IOT [] => TERMINATE 134

# www.cpantesters.org/cpan/report/209fd0b6-e61e-11df-bb29-ad544afd17af
[netbsd]
ABRT [6] => TERMINATE 134
EMT [7] => TERMINATE 135
SEGV [11] => TERMINATE 139
IO [] => IGNORE
IOT [6] = TERMINATE 134

[openbsd]
ILL [4] => TERMINATE 132
ABRT [6] => TERMINATE 134
EMT [7] => TERMINATE 135
BUS [10] => TERMINATE 138
SEGV [11] => TERMINATE 139
IO [] => IGNORE

[mirbsd]
ILL [4] => TERMINATE 132
BUS [10] => TERMINATE 138
SEGV [11] => TERMINATE 139
IO [] => IGNORE

[solaris]
ABRT    [6] => TERMINATE 134
ALRM    [14] => TERMINATE 14
BUS     [10] => TERMINATE 138
CANCEL  [36] => IGNORE
CHLD    [18] => IGNORE
CLD     [18] => IGNORE
CONT    [25] => IGNORE
EMT     [7] => TERMINATE 135
FPE     [8] => TERMINATE 136
FREEZE  [34] => IGNORE
HUP     [1] => TERMINATE 1
ILL     [4] => TERMINATE 132
INT     [2] => TERMINATE 2
IO      [22] => TERMINATE 22
IOT     [6] => TERMINATE 134
JVM1    [39] => IGNORE
JVM2    [40] => IGNORE
KILL    [9] => TERMINATE 9
LOST    [37] => TERMINATE 37
LWP     [33] => IGNORE
NUMxx   [42] => TERMINATE xx
PIPE    [13] => TERMINATE 13
POLL    [22] => TERMINATE 22
PROF    [29] => TERMINATE 29
PWR     [19] => IGNORE
QUIT    [3] => TERMINATE 131
RTMAX   [48] => TERMINATE 48
RTMIN   [41] => TERMINATE 41
SEGV    [11] => TERMINATE 139
STOP    [23] => SUSPEND
SYS     [12] => TERMINATE 140
TERM    [15] => TERMINATE 15
THAW    [35] => IGNORE
TRAP    [5] => TERMINATE 5
TSTP    [24] => SUSPEND
TTIN    [26] => SUSPEND
TTOU    [27] => SUSPEND
URG     [21] => IGNORE
USR1    [16] => TERMINATE 16
USR2    [17] => TERMINATE 17
VTALRM  [28] => TERMINATE 28
WAITING [32] => IGNORE
WINCH   [20] => IGNORE
XCPU    [30] => TERMINATE 30
XFSZ    [31] => TERMINATE 159
XRES    [38] => IGNORE
ZERO    [] => IGNORE
__DIE__ [] => IGNORE
__WARN__ [] => IGNORE
